x-pack/plugins/maps/server/maps_telemetry/maps_telemetry.ts

```diff
@@ -82,6 +94,112 @@ function getIndexPatternsWithGeoFieldCount(indexPatterns: IIndexPattern[]) {
   };
 }
 
+function getEMSLayerCount(layerLists: LayerDescriptor[][]): ILayerTypeCount[] {
+  return layerLists.map((layerList: LayerDescriptor[]) => {
+    const countsById = _(layerList).countBy((layer: LayerDescriptor) => {
+      const isEmsFile =
+        layer.sourceDescriptor !== null
+          ? layer.sourceDescriptor.type === SOURCE_TYPES.EMS_FILE
+          : false;
+      const id = (layer.sourceDescriptor as AbstractSourceDescriptor).id;
+      return isEmsFile && id ? id : 'false';
+    });
+
+    const emsCountsById = countsById.pickBy((val, key) => key !== 'false');
+    const layerTypeCount = emsCountsById.value();
+    return layerTypeCount as ILayerTypeCount;
+  }) as ILayerTypeCount[];
+}
+
+function isGeoshapeIndexPattern(
+  indexPatterns: IIndexPattern[],
+  indexPatternId: string,
+  geoField: string | undefined
+): boolean {
+  if (!geoField) {
+    return false;
+  }
+
+  const matchIndexPattern = indexPatterns.find((indexPattern: IIndexPattern) => {
+    return indexPattern.id === indexPatternId;
+  });
+
+  if (!matchIndexPattern) {
+    return false;
+  }
+
+  const fieldList: IFieldType[] =
+    matchIndexPattern.attributes && matchIndexPattern.attributes.fields
+      ? JSON.parse(matchIndexPattern.attributes.fields)
+      : [];
+
+  const matchField = fieldList.find((field: IFieldType) => {
+    return field.name === geoField;
+  });
+
+  return !!matchField && matchField.type === ES_GEO_FIELD_TYPE.GEO_SHAPE;
+}
+
+function isGeoShapeAggLayer(indexPatterns: IIndexPattern[], layer: LayerDescriptor): boolean {
+  if (layer.sourceDescriptor === null) {
+    return false;
+  }
```

Comment:
Probably good to return early if no `geoField` here instead of on lines 119-121. You could revise this to:
```
const sourceDescriptor: SourceDescriptor = layer.sourceDescriptor;
if (!sourceDescriptor || !sourceDescriptor.geoField) {
    return false;
}
```

Reply:
I'd prefer not too, since we then lose typing-clarity and end-up duck typing (which now we have TS there is no reason to. The TS-typecheck will fail with this pattern too). `geoField` is very specific to ES-sources, and that type-check appears higher-up.